data_diff/database.py

```diff
@@ -714,6 +716,65 @@ def normalize_value_by_type(self, value: str, coltype: ColType) -> str:
 HELP_PRESTO_URI_FORMAT = "presto://<user>@<host>/<catalog>/<schema>"
 
 
+@dataclass
+class MatchUriPath:
+    database_cls: type
+    params: List[str]
+    kwparams: List[str] = []
+    help_str: str
+
+    def match_path(self, dsn):
+        dsn_dict = dict(dsn.query)
+        matches = {}
+        for param, arg in zip_longest(self.params, dsn.paths):
+            if param is None:
+                raise ValueError(f"Too many parts to path. Expected format: {self.help_str}")
+
+            optional = param.endswith('?')
+            param = param.rstrip('?')
+
+            if arg is None:
+                try:
+                    arg = dsn_dict.pop(param)
+                except KeyError:
+                    if not optional:
+                        raise ValueError(f"URI must specify '{param}'. Expected format: {self.help_str}")
+
+                    arg = None
+
+            assert param and param not in matches
+            matches[param] = arg
+
+        for param in self.kwparams:
+            try:
+                arg = dsn_dict.pop(param)
+            except KeyError:
+                raise ValueError(f"URI must specify '{param}'. Expected format: {self.help_str}")
+
+            assert param and arg and param not in matches, (param, arg, matches.keys())
+            matches[param] = arg
+
+        for param, value in dsn_dict.items():
+            if param in matches:
+                raise ValueError(f"Parameter '{param}' already provided as positional argument. Expected format: {self.help_str}")
+
+            matches[param] = value
+
+        return matches
+
+
+MATCH_URI_PATH = {
+    "postgres": MatchUriPath(Postgres, ['database?'], help_str = "postgres://<user>:<pass>@<host>/<database>"),
+    "mysql": MatchUriPath(MySQL, ['database?'], help_str = "mysql://<user>:<pass>@<host>/<database>"),
+    "oracle": MatchUriPath(Oracle, ['database?'], help_str = "oracle://<user>:<pass>@<host>/<database>"),
+    "mssql": MatchUriPath(MsSQL, ['database?'], help_str = "mssql://<user>:<pass>@<host>/<database>"),
+    "redshift": MatchUriPath(Redshift, ['database?'], help_str = "redshift://<user>:<pass>@<host>/<database>"),
+    "snowflake": MatchUriPath(Snowflake, ['database', 'schema'], ['warehouse'], help_str = HELP_SNOWFLAKE_URI_FORMAT),
+    "presto": MatchUriPath(Presto, ['catalog', 'schema'], help_str = HELP_PRESTO_URI_FORMAT),
```

Comment:
Why are some constants, and some inlined?

Reply:
No reason. I can make them all inlined.